Add a public API to expose allocation reports

nical opened the pull request:
This PR adds a public API that exposes enough information for users to build something like the block visualizer without the egui integration. It also adds an offset field in the allocation report.

I would like to integrate (and expose in some form) this in wgpu to debug memory leaks.

The report could expose more information (for example properties of the memory and stack traces) and I haven't added any documentation yet, but it's a good basis for discussing whether the project is OK with exposing something like this and whether the approach is good.

Jasper-Bekkers commented:
Tracks towards #203 so very welcome.

manon-traverse commented on src/d3d12/mod.rs:
```diff
@@ -1099,6 +1099,33 @@ impl Allocator {
         }
     }
+
+    pub fn get_report(&self) -> AllocatorReport {
```
Since this function does more than just get a report I think the name is a bit misleading. Something like `build_report`, `construct_report`, `generate_report` would be better fitting.

nical replied:
I renamed it into `generate_report`

manon-traverse reviewed (commented):
This is a really nice change!
The only part I dislike is the name for creating the report.

MarijnS95 commented on src/d3d12/mod.rs:
```diff
@@ -1099,6 +1100,33 @@ impl Allocator {
         }
     }
+
+    pub fn genetate_report(&self) -> AllocatorReport {
```
```suggestion
    pub fn generate_report(&self) -> AllocatorReport {
```

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -30,13 +30,28 @@ impl AllocationType {
+#[derive(Clone)]
+pub struct MemoryBlockReport {
+    pub size: u64,
+    pub allocations: Range<usize>,
```
Could we have some doc comments explaining how this, for a block, specifies the range in ``[`AllocatorReport::allocations`]`` where all the allocations within this block are found?

MarijnS95 commented on src/d3d12/mod.rs:
```diff
@@ -4,6 +4,7 @@ use std::{backtrace::Backtrace, fmt, sync::Arc};
 
 use log::{debug, warn, Level};
 
+use crate::allocator::AllocatorReport;
```
It looks like the grouped `use` block for this is somewhat below at line 91. No clue why it isn't at the top :sweat_smile:

MarijnS95 commented on src/vulkan/mod.rs:
```diff
@@ -972,6 +972,33 @@ impl Allocator {
+                total_reserved_bytes += block.size;
+                let first_allocation = allocations.len();
+                allocations.extend(block.sub_allocator.report_allocations());
+                blocks.push(MemoryBlockReport {
```
Is it not useful to you yet to report what memory type/heap this block was allocated on?

nical replied:
Absolutely! I was being cautious about exposing internals but I'll take every bit of useful information you are confortable with exposing.

MarijnS95 reviewed (approved):
Nice stuff, as discussed in the `wgpu` Matrix! This would definitely help us long-term with generalizing the visualizer completely, after adding more flags like the relevant heap and other backend-specific items in a somewhat generic way :sweat_smile: 

Would love this to have some doc-comments from the start though, given that -for example- allocations are stored globally instead of per-block and findable via a stored `Range`. Maybe `derive(Debug)` would help as well, but that might be complicated for the top-level `AllocatorReport`.

nical commented:
I made a few notable changes:
 - Added some doc comments
 - reexported `AllocatorReport` and friends at the crate level so that it's actually public
 - Moved the debug implementation of the two allocators into the report (allocators debug impl now just call `generate_report().fmt(f)`).

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,79 @@ impl AllocationType {
+
+        writeln!(
+            f,
+            "================================================================",
```
I personally don't really see the benefit in stdout-like formatting inside `Debug` implementations, especially when they end up getting nested inside other `Debug`able structures.

nical replied:
Fair point. I changed the way that the Debug impl outputs works so that it provides similar information while still working in a nested Debug impl. I suspect that this can turn into its own rabbit hole so perhaps further improvements could be considered in a followups?

It prints something like (without and with pretty printing):
```
AllocatorReport { summary: 16.02 MB / 32.00 MB, blocks: 1, allocations: 2, largest: ["Terrain vertices": 16.00 MB, "Icon texture": 16.00 KB] }
AllocatorReport {
    summary: 16.02 MB / 32.00 MB,
    blocks: 1,
    allocations: 2,
    largest: [
        "Terrain vertices": 16.00 MB,
        "Icon texture": 16.00 KB,
    ],
}
```

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,79 @@ impl AllocationType {
+    pub allocations: Range<usize>,
+}
+
+/// A report that can be generated for debugging purposes using `Allocator::generate_report`.
```
```suggestion
/// A report that can be generated for debugging purposes using `Allocator::generate_report()`.
```

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,79 @@ impl AllocationType {
-    pub(crate) name: String,
-    pub(crate) size: u64,
+pub struct AllocationReport {
+    /// The name provided to the `allocate` function.
```
```suggestion
    /// The name provided to the `allocate()` function.
```

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,79 @@ impl AllocationType {
+    /// All memory blocks.
+    pub blocks: Vec<MemoryBlockReport>,
+    pub total_allocated_bytes: u64,
+    pub total_reserved_bytes: u64,
```
Sum of all blocks (link), including unallocated regions?

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,79 @@ impl AllocationType {
+    pub allocations: Vec<AllocationReport>,
+    /// All memory blocks.
+    pub blocks: Vec<MemoryBlockReport>,
+    pub total_allocated_bytes: u64,
```
Sum of bytes used in ``[`Self::allocations`]`` or something like that?

MarijnS95 reviewed (commented):
Nice stuff! I'm not sure if the original `Debug` format is an appropriate text output for this purpose (maybe better suited for `Display`) but hey, at least it is deduplicated now 🚀

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,73 @@ impl AllocationType {
+/// A report that can be generated for debugging purposes using `Allocator::generate_report()`.
+#[derive(Clone)]
+pub struct AllocatorReport {
+    /// All live allocations.
+    pub allocations: Vec<AllocationReport>,
+    /// All memory blocks.
+    pub blocks: Vec<MemoryBlockReport>,
```
Nit: should we make it a bit more clear to the caller what the difference is between a block (larger chunk of memory reserved by `gpu-allocator` to hand out to users) versus `allocations` (actual pieces of memory within these blocks, requested by the user)?

Right now this explanation lives on the fields below.

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,73 @@ impl AllocationType {
+    pub allocations: Range<usize>,
+}
+
+/// A report that can be generated for debugging purposes using `Allocator::generate_report()`.
```
I'd say "informational purposes". Statistics are generally useful, not just for debugging.

MarijnS95 commented on src/allocator/mod.rs:
```diff
@@ -29,20 +29,73 @@ impl AllocationType {
+
+impl fmt::Debug for AllocatorReport {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut allocations = self.allocations.clone();
+        allocations.sort_by_key(|alloc| std::cmp::Reverse(alloc.size));
+
+        let max_num_allocations_to_print = f.precision().unwrap_or(usize::MAX);
+        allocations.truncate(max_num_allocations_to_print);
```
I wonder if Rust has a way to only collect+sort the highest `N` number of items. This list might potentially be large and sorting a lot of small allocations when you only care about a few large ones seems a bit of a waste :)

nical commented:
I added metal and made a few other adjustments from the review comments. I still haven't added information about the memory blocks to the report. To do that I need to spend a bit of time looking at each backend, finding the common denominators and storing the right info in the right places for the report to pick up. It's not much but because I am currently juggling a lot of tasks on different projects I'm having some difficulty finding the time to wrap my head around that bit.

Would you be OK with landing the current feature set?
Should I bump the patch version in this PR?

MarijnS95 commented:
> Should I bump the patch version in this PR?

I'm sure `wgpu` has special release processes that don't like arbitrary PRs bumping the version either :wink: 

---

We currently have a CI failure related to the whole `metal-rs` "debacle" that might be blocking the merge of this PR. We'll have to resolve that first before this can be merged.

MarijnS95 commented:
> I don't think there's a specific rule about that but since wgpu's release process is fair bit more involved than most crates, it indeed tends to be done in isolation.


So is ours, and I think for every crate it's a much nicer policy to have an explicit "Release XXX" commit in their history rather than doing it within a feature PR.

MarijnS95 commented:
@nical would you mind to sign your commits?

MarijnS95 commented:
@nical JFYI, despite being subscribed to this repo and PR I don't receive any notifications for your pushes, and just happened to see the update at the right moment by accident. Please ping me once I have to approve the CI again, since GitHub delegates their anti-CI-abuse measures to repository owners and maintainers :sweat_smile:

nical commented:
Good to know, @MarijnS95. I just pushed another commit to fix the fmt issue
